src/executor/enter.rs

```diff
@@ -0,0 +1,89 @@
+use std::prelude::v1::*;
+use std::cell::{RefCell, Cell};
+use std::fmt;
+
+thread_local!(static ENTERED: Cell<bool> = Cell::new(false));
+
+/// Represents an executor context.
+///
+/// For more details, see [`enter` documentation](fn.enter.html)
+pub struct Enter {
+    on_exit: RefCell<Vec<Box<Callback>>>,
+    permanent: bool,
+}
+
+/// Marks the current thread as being within the dynamic extent of an
+/// executor.
+///
+/// Executor implementations should call this function before blocking the
+/// thread. If `None` is returned, the executor should fail by panicking or
+/// taking some other action without blocking the current thread. This prevents
+/// deadlocks due to multiple executors competing for the same thread.
+///
+/// # Panics
+///
+/// Panics if the current thread is *already* marked.
+pub fn enter() -> Option<Enter> {
+    ENTERED.with(|c| {
+        if c.get() {
+            None
+        } else {
+            c.set(true);
+
+            Some(Enter {
+                on_exit: RefCell::new(Vec::new()),
+                permanent: false,
+            })
+        }
+    })
+}
+
+impl Enter {
+    /// Register a callback to be invoked if and when the thread
+    /// ceased to act as an executor.
+    pub fn on_exit<F>(&self, f: F) where F: FnOnce() + 'static {
+        self.on_exit.borrow_mut().push(Box::new(f));
+    }
+
+    /// Treat the remainder of execution on this thread as part of an
+    /// executor; used mostly for thread pool worker threads.
+    ///
+    /// All registered `on_exit` callbacks are *dropped* without being
+    /// invoked.
+    pub fn make_permanent(mut self) {
+        self.permanent = true;
```

Comment:
This needs to actually drop the callbacks (and no further ones should be accepted).

Reply:
Isn't that handled by the fact that it takes `self` via ownership?

Reply:
Missed that. You're right.